### PR TITLE
[CodeCamp #83] Support Restormer model

### PR DESCRIPTION
**Pull request by AlexZou14**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation
Support Restormer of image restoration
Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

**Before PR**:

- [ ] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmediting/blob/master/.github/CONTRIBUTING.md) to create this PR.
- [ ] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmediting/blob/master/.github/CONTRIBUTING.md) are used to fix the potential lint issues.
- [ ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmedit/models/editors/restormer/restormer_net.py [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+
+def to_3d(x):
+    return rearrange(x, 'b c h w -> b (h w) c')
+
+
+def to_4d(x, h, w):
+    return rearrange(x, 'b (h w) c -> b c h w', h=h, w=w)
+
+
+class BiasFree_LayerNorm(BaseModule):
```
Comment (Z-Fran) [inline mmedit/models/editors/restormer/restormer_net.py:22]: Add docstrings for every class and every forward function. You can reference other models.

File: configs/_base_/datasets/deblurring_test_config.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    dataset=dict(
+        type='BasicImageDataset',
+        metainfo=dict(dataset_type='GoPro', task_name='deblurring'),
+        data_root=gopro_data_root,
+        data_prefix=dict(img='input', gt='target'),
+        pipeline=test_pipeline))
+gopro_evaluator = [
+    dict(type='MAE', prefix='GoPro'),
+    dict(type='PSNR', prefix='GoPro'),
+    dict(type='SSIM', convert_to='Y', prefix='GoPro'),
```
Comment (Z-Fran) [inline configs/_base_/datasets/deblurring_test_config.py]: why test SSIM in 'y' channel?

File: configs/_base_/datasets/deblurring_test_config.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    drop_last=False,
+    sampler=dict(type='DefaultSampler', shuffle=False),
+    dataset=dict(
+        type='BasicImageDataset',
+        metainfo=dict(dataset_type='RealBlur_R', task_name='deblurring'),
+        data_root=realblurr_data_root,
+        data_prefix=dict(img='input', gt='target'),
+        pipeline=test_pipeline))
+realblurr_evaluator = [
+    dict(type='MAE', prefix='RealBlurR'),
```
Comment (Z-Fran) [inline configs/_base_/datasets/deblurring_test_config.py]: remove MAE

File: configs/_base_/datasets/deraining_test_config.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    num_workers=4,
+    persistent_workers=False,
+    drop_last=False,
+    sampler=dict(type='DefaultSampler', shuffle=False),
+    dataset=dict(
+        type='BasicImageDataset',
+        metainfo=dict(dataset_type='Rain100L', task_name='deraining'),
+        data_root=rain100l_data_root,
+        data_prefix=dict(img='input', gt='target'),
+        # ann_file='meta_info_rain100l_test.txt',
```
Comment (Z-Fran) [inline configs/_base_/datasets/deraining_test_config.py:46]: remove this line

File: configs/_base_/datasets/deraining_test_config.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    num_workers=4,
+    persistent_workers=False,
+    drop_last=False,
+    sampler=dict(type='DefaultSampler', shuffle=False),
+    dataset=dict(
+        type='BasicImageDataset',
+        metainfo=dict(dataset_type='Test100', task_name='deraining'),
+        data_root=test100_data_root,
+        data_prefix=dict(img='input', gt='target'),
+        # ann_file='meta_info_test100_test.txt',
```
Comment (Z-Fran) [inline configs/_base_/datasets/deraining_test_config.py:65]: remove this line

File: configs/_base_/datasets/deraining_test_config.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    num_workers=4,
+    persistent_workers=False,
+    drop_last=False,
+    sampler=dict(type='DefaultSampler', shuffle=False),
+    dataset=dict(
+        type='BasicImageDataset',
+        metainfo=dict(dataset_type='Test1200', task_name='deraining'),
+        data_root=test1200_data_root,
+        data_prefix=dict(img='input', gt='target'),
+        # ann_file='meta_info_test1200_test.txt',
```
Comment (Z-Fran) [inline configs/_base_/datasets/deraining_test_config.py:84]: remove this line

File: configs/_base_/datasets/deraining_test_config.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+    num_workers=4,
+    persistent_workers=False,
+    drop_last=False,
+    sampler=dict(type='DefaultSampler', shuffle=False),
+    dataset=dict(
+        type='BasicImageDataset',
+        metainfo=dict(dataset_type='Test2800', task_name='deraining'),
+        data_root=test2800_data_root,
+        data_prefix=dict(img='input', gt='target'),
+        # ann_file='meta_info_test2800_test.txt',
```
Comment (Z-Fran) [inline configs/_base_/datasets/deraining_test_config.py:103]: remove this line

File: configs/restormer/restormer_official_gopro.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
+_base_ = [
+    '../_base_/default_runtime.py',
+    '../_base_/datasets/defocus_deblurring_test_config.py'
```
Comment (Z-Fran) [inline configs/restormer/restormer_official_gopro.py:3]: this should be deblurring_test_config.py?

File: configs/restormer/restormer_official_dpdd-single.py [hunk 9]
```diff
@@ -0,0 +1,30 @@
+_base_ = [
+    '../_base_/default_runtime.py',
+    '../_base_/datasets/defocus_deblurring_test_config.py'
```
Comment (Z-Fran) [inline configs/restormer/restormer_official_dpdd-single.py:3]: rename

File: mmedit/models/editors/restormer/restormer_net.py [hunk 10]
```diff
@@ -0,0 +1,527 @@
...
+    """Reshape input tensor."""
+    return rearrange(x, 'b c h w -> b (h w) c')
+
+
+def to_4d(x, h, w):
+    """Reshape input tensor."""
+    return rearrange(x, 'b (h w) c -> b c h w', h=h, w=w)
+
+
+class BiasFree_LayerNorm(BaseModule):
```
Comment (Z-Fran) [inline mmedit/models/editors/restormer/restormer_net.py:24]: Add a description

File: mmedit/models/editors/restormer/restormer_net.py [hunk 11]
```diff
@@ -0,0 +1,527 @@
...
+        if isinstance(normalized_shape, numbers.Integral):
+            normalized_shape = (normalized_shape, )
+        normalized_shape = torch.Size(normalized_shape)
+
+        assert len(normalized_shape) == 1
+
+        self.weight = nn.Parameter(torch.ones(normalized_shape))
+        self.normalized_shape = normalized_shape
+
+    def forward(self, x):
```
Comment (Z-Fran) [inline mmedit/models/editors/restormer/restormer_net.py:41]: Add docstring

File: mmedit/models/editors/restormer/restormer_net.py [hunk 12]
```diff
@@ -0,0 +1,527 @@
...
+        sigma = x.var(-1, keepdim=True, unbiased=False)
+        return x / torch.sqrt(sigma + 1e-5) * self.weight
+
+
+class WithBias_LayerNorm(BaseModule):
```
Comment (Z-Fran) [inline mmedit/models/editors/restormer/restormer_net.py:46]: Add a description

File: mmedit/models/editors/restormer/restormer_net.py [hunk 13]
```diff
@@ -0,0 +1,527 @@
...
+            normalized_shape = (normalized_shape, )
+        normalized_shape = torch.Size(normalized_shape)
+
+        assert len(normalized_shape) == 1
+
+        self.weight = nn.Parameter(torch.ones(normalized_shape))
+        self.bias = nn.Parameter(torch.zeros(normalized_shape))
+        self.normalized_shape = normalized_shape
+
+    def forward(self, x):
```
Comment (Z-Fran) [inline mmedit/models/editors/restormer/restormer_net.py:64]: Add docstring

File: mmedit/models/editors/restormer/restormer_net.py [hunk 14]
```diff
@@ -0,0 +1,527 @@
...
+        mu = x.mean(-1, keepdim=True)
+        sigma = x.var(-1, keepdim=True, unbiased=False)
+        return (x - mu) / torch.sqrt(sigma + 1e-5) * self.weight + self.bias
+
+
+class LayerNorm(BaseModule):
```
Comment (Z-Fran) [inline mmedit/models/editors/restormer/restormer_net.py:70]: Add a description

File: mmedit/models/editors/restormer/restormer_net.py [hunk 15]
```diff
@@ -0,0 +1,527 @@
...
+    """
+
+    def __init__(self, dim, LayerNorm_type):
+        super(LayerNorm, self).__init__()
+        if LayerNorm_type == 'BiasFree':
+            self.body = BiasFree_LayerNorm(dim)
+        else:
+            self.body = WithBias_LayerNorm(dim)
+
+    def forward(self, x):
```
Comment (Z-Fran) [inline mmedit/models/editors/restormer/restormer_net.py:85]: Add docstring

File: tests/test_models/test_editors/test_restormer/test_restormer_net.py [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+        bias=False,
+        LayerNorm_type='WithBias',
+        dual_pixel_task=True)
+    img = torch.rand(1, 6, 64, 64)
+    output = net(img)
+    assert isinstance(output, torch.Tensor)
+    assert output.shape == (1, 3, 64, 64)
+
+
+def test_swinir_cuda():
```
Comment (Z-Fran) [inline tests/test_models/test_editors/test_restormer/test_restormer_net.py:77]: typo

File: tests/test_models/test_editors/test_restormer/test_restormer_net.py [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+        num_blocks=[4, 6, 6, 8],
+        num_refinement_blocks=4,
+        heads=[1, 2, 4, 8],
+        ffn_expansion_factor=2.66,
+        bias=False,
+        LayerNorm_type='WithBias',
+        dual_pixel_task=False)
+    img = torch.rand(1, 3, 64, 64)
+
+    # x4 model lightweight SR forward (gpu)
```
Comment (Z-Fran) [inline tests/test_models/test_editors/test_restormer/test_restormer_net.py:91]: typo

**Review by Z-Fran — APPROVED** [review]

**Review by Z-Fran — APPROVED** [review]

**Review by plyfager — APPROVED** [review]

**Pull request merged**